sdk/eventhub/event-hubs/src/eventHubSender.ts

```diff
@@ -526,6 +516,113 @@ export class EventHubSender extends LinkEntity {
+            this._context.connectionId,
+            this.name
+          );
+          resolve();
+        };
+        onRejected = (context: EventContext) => {
+          removeListeners();
+          log.error("[%s] Sender '%s', got event rejected.", this._context.connectionId, this.name);
+          const err = translate(context!.delivery!.remote_state!.error);
+          log.error(err);
+          reject(err);
+        };
+        onReleased = (context: EventContext) => {
+          removeListeners();
+          log.error("[%s] Sender '%s', got event released.", this._context.connectionId, this.name);
+          let err: Error;
+          if (context!.delivery!.remote_state!.error) {
+            err = translate(context!.delivery!.remote_state!.error);
+          } else {
+            err = new Error(
+              `[${this._context.connectionId}] Sender '${this.name}', ` +
+                `received a release disposition.Hence we are rejecting the promise.`
+            );
+          }
+          log.error(err);
+          reject(err);
+        };
+        onModified = (context: EventContext) => {
+          removeListeners();
+          log.error("[%s] Sender '%s', got event modified.", this._context.connectionId, this.name);
+          let err: Error;
+          if (context!.delivery!.remote_state!.error) {
+            err = translate(context!.delivery!.remote_state!.error);
+          } else {
+            err = new Error(
+              `[${this._context.connectionId}] Sender "${this.name}", ` +
+                `received a modified disposition.Hence we are rejecting the promise.`
+            );
+          }
+          log.error(err);
+          reject(err);
+        };
+
+        const removeListeners = (): void => {
+          clearTimeout(waitTimer);
+          // When `removeListeners` is called on timeout, the sender might be closed and cleared
+          // So, check if it exists, before removing listeners from it.
+          if (abortSignal) {
+            abortSignal.removeEventListener("abort", onAborted);
+          }
+          if (this._sender) {
+            this._sender.removeListener(SenderEvents.rejected, onRejected);
+            this._sender.removeListener(SenderEvents.accepted, onAccepted);
+            this._sender.removeListener(SenderEvents.released, onReleased);
+            this._sender.removeListener(SenderEvents.modified, onModified);
+          }
+        };
+
+        const actionAfterTimeout = () => {
+          removeListeners();
+          const desc: string =
+            `[${this._context.connectionId}] Sender "${this.name}" with ` +
+            `address "${this.address}", was not able to send the message right now, due ` +
+            `to operation timeout.`;
+          log.error(desc);
+          const e: Error = {
+            name: "OperationTimeoutError",
+            message: desc
+          };
+          return reject(translate(e));
+        };
+
+        waitTimer = setTimeout(
```

Comment:
I think you probably could have just moved this section to the top of the operation. The `actionAfterTimeout` wouldn't be invoked during this tick of the event-loop, so everything it calls would be defined when it was finally invoked.